### PR TITLE
Use state before join to determine if we `_should_perform_remote_join`

### LINKED ISSUE #13269: `_should_perform_remote_join` inspects the current state of the room, not the state of the would-be join
In normal operation (<= 10 fwd extremities, no races over the relevant state) the two state maps would be one and the same. But they need not be. For consistency's sake we should run any state-dependent checks against the state prior to the would-be join. That's what other HSes will do!

Additionally, the current state of the room is cached in-application and therefore subject to cache invalidation races. By using the state at the would-be join we avoid this race. This explains a particularly gnarly complement flake (xref #13161):

1. Alice invites Bob to a restricted room that Bob would otherwise be unable to join.
2. Bob tries to join the room. We select `prev_events` for Bob's join; resolve across them to produce the state before this join; and determine that his membership is `invite`.
3. Bob's HS calls `_should_perform_remote_join` to work out if it can needs to use a remote `make_join`/`send_join` handshake. To do so, it requests the current state of the room. __This is cached and subject to a cache invalidation race;__ the cached current state may not have been updated to include the fact that Bob is invited.

### PR DESCRIPTION
**Pull request by DMRobertson**
Before this change, we were using the current state of the room. (The two are usually but not always consistent.)

Fixes #13269, see there for context and motivation. Commitwise review recommended.

#10254 introduced some of the logic I'm changing here.

### CODE REVIEW

File: synapse/state/__init__.py [hunk 1]
```diff
@@ -137,22 +137,27 @@ def __init__(self, hs: "HomeServer"):
...
 
         Args:
-            room_id:
-            latest_event_ids: The forward extremities to resolve.
+            room_id: the room_id containing the given events.
+            event_ids: the events whose state should be fetched and resolved.
 
         Returns:
-            the state dict, mapping from (event_type, state_key) -> event_id
+            the resolution of the states after the given event IDs. This is a mapping
```
Comment (H-Shay) [inline synapse/state/__init__.py:156]: I have found the term "state dict" useful in the past-would it be amenable to keep it in along with your change (which is edifying), ie something like "the state dict which is a  resolution..[the rest of your addition here]"

**Comment by DMRobertson** [comment]
Thanks Shay, you're absolutely right. Let me change the title.

File: synapse/state/__init__.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/state/__init__.py]: How do you find [`6f6140e` (#13270)](https://github.com/matrix-org/synapse/pull/13270/commits/6f6140e7f5f9217a945037cc87fa571e10bbcbf4)?

**Comment by DMRobertson** [comment]
@richvdh might be interested in this one, since we talked it over yesterday.

**Comment by H-Shay** [comment]
I have no other comments, I am happy to give a green check but do you want to have vdh look at it?

**Comment by DMRobertson** [comment]
Let's do that---no harm in a second opinion.

File: synapse/state/__init__.py [hunk 2]
```diff
@@ -151,22 +151,27 @@ def __init__(self, hs: "HomeServer"):
...
         Args:
-            room_id:
-            latest_event_ids: The forward extremities to resolve.
+            room_id: the room_id containing the given events.
+            event_ids: the events whose state should be fetched and resolved.
 
         Returns:
-            the state dict, mapping from (event_type, state_key) -> event_id
+            the state dict (a mapping from (event_type, state_key) -> event_id) which
+            holds the resolution of the state dicts after the given event IDs.
```
Comment (richvdh) [inline synapse/state/__init__.py:171]:
```suggestion
            holds the resolution of the states after the given event IDs.
```

File: synapse/state/__init__.py [hunk 3]
```diff
@@ -151,22 +151,27 @@ def __init__(self, hs: "HomeServer"):
...
         """
         logger.debug("calling resolve_state_groups from get_current_state_ids")
```
Comment (richvdh) [inline synapse/state/__init__.py:173]: this is no longer accurate. update please!

**Review by richvdh — APPROVED** [review]
nitty nits, but lgtm, thanks!

**Pull request merged**